### PR TITLE
Add Mask R-CNN

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the classic Mask R-CNN framework for object detection and instance segmentation.

To do/to be discussed:

- [ ] where to place utilities like NMS, loss computation, samplers
- [ ] whether to create dummies for torchvision-backed models
- [ ] how to add support for the object detection pipeline - either add `**kwargs` to each `post_process_object_detection` method, or add specific logic for Mask R-CNN inside `object_detection_pipeline.py`

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
The PR is not ready for review yet. The code is not readable and makes heavy use of one-letter variables which makes it impossible to read and maintain. None of the docstring follow the Hugging Face format and as such have been completely destroyed by the formatter. There are multiple asserts everyewhere. The classes should use camel-cased names, so be prefixed by `MaskRcnn`.

For your initial questions, utils such as nms should leave in the model folder and there should be no dummies for torchvision specific objects. The model can have a soft dependency on torchvision by raising an error at init if torchvision is not installed.

**Comment by amyeroberts** [comment]
@NielsRogge As @sgugger mentions, the PR is still in WIP state. Happy to review once transformers ready :)

**Comment by NielsRogge** [comment]
I've updated all docstrings and variable names, PR is ready for another review

**Review by amyeroberts — COMMENTED** [review]
@NielsRogge I started reviewing, but this PR is still not quite in a ready state. In particular, there's a few things to address:
* TODOs left all over the code
* Lots of remaining commented out code here and there
* Many important methods and functions still don't have doc strings, and some aren't complete e.g. for the configuration class. Detailed descriptions of what methods are doing would make reviewing this a lot easier. 
* Unclear variable names in code and docstrings e.g. lots of x1, y1, N. The SAM processing code is a good example of making these clearer for bboxes. 
* In some of the model-side processing code, there's switching to CPU and casting back and forth between torch and numpy - it's not clear why. 

Some general comments: 
* In terms of the code in general, it's quite hard to follow and read. The many additional mask RCNN modules should be moved into the modeling file. Some of the extra classes have logic which I don't think is necessary for this implementation e.g. all of the specific `__repr__` and `__nice__` methods. We can remove these and make things a bit tighter. 
* `unmap` I think I've seen before :) It's not very clear what 'unmapping' means - the method could do with a rename. 
* Let's try to avoid name clashes with common ML classes e.g. `CrossEntropyLoss`
* One question I have is about the NMS implementation - is there a reason for having this rather than use e.g. torchvision's?
* There's a bit of if/else logic for onnx conversions which makes it harder to read. If the logic is equivalent, let's just use the onnx compatible logic directly 
* Some layers can be reworked to just take the config file or reduce kwargs e.g. many of the instance attributes in `MaskRCNNSingleRoIExtractor` aren't used.

I know this is a big model to implement, and you're very busy. Let us know if you'd like someone to take over this PR. As the architecture is complex and doesn't fit easily into the 'transformers' library default patterns, another alternative would be hosting the model code directly on the hub first.

**Comment by aymanechilah** [comment]
Hello @NielsRogge, what is the status of this? Can we use it? I'm very interested...

**Comment by bit-scientist** [comment]
Hi, @NielsRogge looking forward to it. Could you, for now, recommend a robust text detector available here to combine with TrOCR. I would like to see how well the two work with the help of HF🤗.
